pegen/grammar_visualizer.py

```diff
@@ -0,0 +1,92 @@
+import argparse
+import tokenize
+import enum
+
+from pegen.tokenizer import Tokenizer
+from pegen.tokenizer import grammar_tokenizer
+from pegen.grammar2 import GrammarParser
+
+
+argparser = argparse.ArgumentParser(prog='pegen', description="Experimental PEG-like parser generator")
+argparser.add_argument('filename', help="Grammar description")
+
+
+class Direction(enum.Enum):
+        UP = 0
+        DOWN = 1
+        NEUTRAL = 2
+        NONE = 3
+
+
+class GrammarPrinter:
+
+    def children(self, node):
+        for value in node:
+            if isinstance(value, list):
+                yield from value
+            else:
+                yield value
+    def nb_children(self, node):
+        return sum(self.nb_children(child) for child in self.children(node)) + 1
+
+    def name(self, node):
+        if (name := getattr(node, "name", None)):
+            return node.__class__.__name__ + f"({node.name})"
+        elif (value := getattr(node, "value", None)):
+            return node.__class__.__name__ + f"({node.value})"
+        else:
+            return node.__class__.__name__
+
+
+    def print_grammar(self, current_node, indent='', last=Direction.NEUTRAL, printer=print):
+
+        child_sizes = {child: self.nb_children(child) for child in self.children(current_node)}
+
+        up = sorted(child_sizes, key=lambda key: child_sizes[key])
+
+        down = []
+        while up and sum(child_sizes[node] for node in down) < sum(child_sizes[node] for node in up):
+            down.append(up.pop())
+
+        for child in up:
+            next_last = Direction.UP if up.index(child) == 0 else Direction.NONE
+            next_indent = (f"{indent}{' ' if last == Direction.UP else '│'}"
+                           f"{' ' * len(self.name(current_node))}")
+            self.print_grammar(child, next_indent, next_last)
+
+        if last == Direction.UP:
```

Comment:
These two if/elif blocks look low in density. Maybe the first can use a pre-populated dict? The second could be an if-expression.